django/db/models/sql/query.py

```diff
@@ -1544,9 +1544,7 @@ def build_filter(
             if len(targets) == 1:
                 col = self._get_col(targets[0], join_info.final_field, alias)
             else:
-                col = MultiColSource(
-                    alias, targets, join_info.targets, join_info.final_field
-                )
+                col = Cols(alias, targets, join_info.targets, join_info.final_field)
```

Comment:
Do you know under which circumstances `targets != join_info.targets`? This seems to be the only usage where `sources` and `targets` might be different. If it never happens we could likely greatly simplify `Cols`.

Reply:
Yes, this is used by `ForeignObject`, where `sources = from_fields` and `targets = to_fields`. So here it will always be different.

When Cols is created by the composite field, it will always be the same.

This does sound like we shouldn't combine MultiColSource and Cols. At the same time, both expressions resolve into tuple operations, so it would be nice to combine them.

The other solution I thought about is to have 2 classes:
1. MultiColSource = base class, used by ForeignObject
2. Cols(MultiColSource) = sets targets and sources to the same field, used by composite fields

This way in the code we could keep checking if it's an instance of MultiColSource.

Reply:
One thing that eludes here though is that in the case of a single target (the `len(targets) == 1` branch above) returning a single `Col` is enough for the resolving to work fine.

Knowing that, why wouldn't a `Tuple` of calls to `Tuple(self._get_col(target, join_info.final_field, alias) for target in targets)` be enough for the related lookups `isinstance(self.lhs, Tuple)` to take the right decisions? What makes `SubqueryConstraint` so special?

Reply:
It's a great question, I haven't dived deep into `SubqueryConstraint`, let me read into this...

Reply:
`SubqueryConstraint` is a little convoluted.

Whenever doing a related IN lookup on multiple fields, first, the related table (A) is joined. Then, the same table (A as B) is queried in a sub-select.

In MySQL, it does `(A.col1, A.col2) IN (SELECT B.col1, B.col2 FROM A as B)`.
In other backends, it does `EXISTS (SELECT B.col1, B.col2 FROM A as B WHERE A.col1 = B.col1 AND A.col2 = B.col2)`.